src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs

```diff
@@ -659,36 +659,55 @@ public StringBuilder Append(char value, int repeatCount)
                 return this;
             }
 
-            // this is where we can check if the repeatCount will put us over m_MaxCapacity
-            // We are doing the check here to prevent the corruption of the StringBuilder.
-            int newLength = Length + repeatCount;
-            if (newLength > m_MaxCapacity || newLength < repeatCount)
+            char[] chunkChars = m_ChunkChars;
+            int chunkLength = m_ChunkLength;
+
+            if (((uint)chunkLength + (uint)repeatCount) <= (uint)chunkChars.Length)
             {
-                throw new ArgumentOutOfRangeException(nameof(repeatCount), SR.ArgumentOutOfRange_LengthGreaterThanCapacity);
+                chunkChars.AsSpan(chunkLength, repeatCount).Fill(value);
+                m_ChunkLength += repeatCount;
             }
-
-            int index = m_ChunkLength;
-            while (repeatCount > 0)
+            else
             {
-                if (index < m_ChunkChars.Length)
-                {
-                    m_ChunkChars[index++] = value;
-                    --repeatCount;
-                }
-                else
-                {
-                    m_ChunkLength = index;
-                    ExpandByABlock(repeatCount);
-                    Debug.Assert(m_ChunkLength == 0);
-                    index = 0;
-                }
+                AppendWithExpansion(value, repeatCount);
             }
 
-            m_ChunkLength = index;
             AssertInvariants();
             return this;
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void AppendWithExpansion(char value, int repeatCount)
+        {
+            Debug.Assert(repeatCount > 0, "Invalid length; should have been validated by caller.");
+
+            // Check if the repeatCount will put us over m_MaxCapacity
+            if ((uint)(repeatCount + Length) > (uint)m_MaxCapacity)
+            {
+                throw new ArgumentOutOfRangeException(nameof(repeatCount), SR.ArgumentOutOfRange_LengthGreaterThanCapacity);
+            }
+
+            char[] chunkChars = m_ChunkChars;
+            int chunkLength = m_ChunkLength;
+
+            // Fill the rest of the current chunk
+            int firstLength = chunkChars.Length - chunkLength;
+            if (firstLength > 0)
+            {
+                chunkChars.AsSpan(chunkLength, firstLength).Fill(value);
+                m_ChunkLength = chunkChars.Length;
+            }
+
+            // Expand the builder to add another chunk
+            int restLength = repeatCount - firstLength;
```

Comment:
Simpler to just [modify](https://github.com/dotnet/runtime/pull/86287/files#diff-072e8df6b3d249d58da2b688cbd9369ae5d1e9fb4c1c1da77a19384f3f7711a0R703-R708) like:
```csharp
repeatCount -= firstLength;
ExpandByABlock(repeatCount);
Debug.Assert(m_ChunkLength == 0, "A new block was not created.");

// Fill the new chunk with the remaining part of repeatCount
m_ChunkChars.AsSpan(0, repeatCount).Fill(value);
m_ChunkLength = repeatCount;
```

(oh how I wish `Span<T>.Fill()` returned how many `T` it filled... sigh.)

Reply:
The cost of having a new variable here is close to zero, but if reusing repeatCount reads better then I'm all for it